onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,119 @@
+void AddReductionParams(T* params, const std::vector<int64_t>& axes, bool keepdims, bool noop_with_empty_axes) {
+  params->set_keepdims(keepdims);
+
+  for (auto& axis : axes)
+    params->add_axes(axis);
+
+  if (axes.size() == 0 && !noop_with_empty_axes)
+    params->set_reduceall(true);
+}
+} // namespace
+
+void ReductionOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  const auto& input_defs(node.InputDefs());
+
+  // We have already embedded the axes into the CoreML layer.
+  // No need to copy them later to reduce memory consumption.
+  if (input_defs.size() > 1)
+    model_builder.AddInitializerToSkip(input_defs[1]->Name());
+}
+
+Status ReductionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                                 const logging::Logger& /* logger */) const {
+  const auto& op_type(node.OpType());
+  const auto& input_defs(node.InputDefs());
+  const auto& initializers(model_builder.GetInitializerTensors());
+
+  std::vector<int64_t> axes;
+
+  NodeAttrHelper helper(node);
+  if (input_defs.size() > 1 && input_defs[1]->Exists()) {
+    auto& axes_tensor = *initializers.at(input_defs[1]->Name());
+    Initializer axes_initializer(axes_tensor);
+    int64_t* data = axes_initializer.data<int64_t>();
+    int64_t size = axes_initializer.size();
+
+    axes = std::vector<int64_t>(data, data + size);
+  } else if (helper.HasAttr("axes")) {
+    axes = helper.Get("axes", std::vector<int64_t>{});
+  }
+  const bool keepdims = helper.Get("keepdims", 1) != 0;
+  const bool noop_with_empty_axes = helper.Get("noop_with_empty_axes", 0) != 0;
+
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  if (op_type == "ReduceSum") {
+    AddReductionParams(layer->mutable_reducesum(), axes, keepdims, noop_with_empty_axes);
+  } else if (op_type == "ReduceMean") {
+    AddReductionParams(layer->mutable_reducemean(), axes, keepdims, noop_with_empty_axes);
+  } else {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "ReductionOpBuilder::AddToModelBuilderImpl, unknown op: ", op_type);
+  }
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+  return Status::OK();
+}
+#endif
+
+bool ReductionOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                                           const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  if (input_defs.size() > 1 && input_defs[1]->Exists()) {
+    const auto& axes_name = input_defs[1]->Name();
+    const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+    if (!Contains(initializers, axes_name)) {
+      LOGS(logger, VERBOSE) << "Axes of reduction must be a constant initializer";
+      return false;
```

Comment:
The changes look fine to me but I think there's a small testing gap. The existing unit tests for ReduceMean don't seem to have a test that adds `axes` as a input using an initializer. ReduceSum has a couple. e.g. [https://github.com/microsoft/onnxruntime/blob/b20d5e85d5ca2333e9c0f52b595254c41c40f7ab/onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc#L2363](https://github.com/microsoft/onnxruntime/blob/b20d5e85d5ca2333e9c0f52b595254c41c40f7ab/onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc#L2363)

Could you please add a test to ensure we have unit test coverage of the ReduceSum path for recent opsets where `axes` is an input? The ReduceSum test is a good example 
- use AddInput with the `is_initializer` parameter set to `true` 
  - vs. older tests using AddAttribute
- set opset to 18 in the OpTester constructor 
  - axes was moved to an input for ReduceSum in opset 13 and the others in opset 18

Can probably cut-and-paste an existing test or two and just tweak it with those changes.

Reply:
Thanks @skottmckay. I implemented 2 tests for ReduceMean op with axes as an initializer input.